feat: ipfs-http-client -> kubo-rpc-client

SgtPooki opened the pull request:
fixes https://github.com/ipfs/ipfs-webui/issues/2079

Note that this PR includes patches for kubo-rpc-client and ipld-explorer-components, both of which have PRs out:

- [x] https://github.com/ipfs/js-kubo-rpc-client/pull/145
- [ ] https://github.com/ipfs/ipld-explorer-components/pull/356

lidel commented:
UnixFS examples in IPLD Explorer seems to be  broken:
- http://bafybeiamvv2dr2nhvhgyipbobewv6gmjht5rphcru3ptcjdcfogwts6zgu.ipfs.localhost:8080/#/explore/QmSnuWmxptJZdLJpKRarxBMS2Ju2oANVrgbr2xWbie9b2D
- http://bafybeiamvv2dr2nhvhgyipbobewv6gmjht5rphcru3ptcjdcfogwts6zgu.ipfs.localhost:8080/#/explore/QmdmQXB2mzChmMeKY47C43LxUdg1NDJ5MWcKMKxDu7RgQm

Firefox:
> ![2023-02-13-213303_1260x284_scrot](https://user-images.githubusercontent.com/157609/218568544-cb265a5c-727f-45f9-ba7f-871030631277.png)

> ![2023-02-13-213441_1172x230_scrot](https://user-images.githubusercontent.com/157609/218568831-24877b69-b269-47e1-b128-6450bcc22185.png)

DAG-CBOR also stopped working:

http://bafybeiamvv2dr2nhvhgyipbobewv6gmjht5rphcru3ptcjdcfogwts6zgu.ipfs.localhost:8080/#/explore/ipfs/bafyreihnpl7ami7esahkfdnemm6idx4r2n6u3apmtcrxlqwuapgjsciihy produces:

lidel reviewed (changes requested):
(if we were missing tests for these, we should add them as part of this PR, to avoid regressions in the future)

SgtPooki commented on test/e2e/setup/ipfs-backend.js:
```diff
@@ -1,5 +1,4 @@
-// import * as kuboRpcModule from 'kubo-rpc-client'
-import ipfsHttpModule from 'ipfs-http-client'
+import * as kuboRpcModule from 'kubo-rpc-client'
```
I think this is an artifact from addressing import issues in storybook/jest/runtime. I will look at normalizing the imports

SgtPooki replied:
`import * as kuboRpcModule from 'kubo-rpc-client'` allows us to pass the correct object to Ctl.createFactory, while also allowing us to do `kuboRpcModule.create`. if I just imported create, I would have to do:


```js
const factory = Ctl.createFactory({
      kuboRpcModule: { create },
```

which feels worse.

SgtPooki commented on package.json:
```diff
@@ -58,22 +58,22 @@
     "i18next-chained-backend": "^3.0.2",
     "i18next-http-backend": "^1.3.2",
-    "i18next-icu": "^2.0.3",
+    "i18next-icu": "2.0.3",
```
so.. fun story: we have dependency conflicts when i18next-icu: ^2.0.3 resolves to 2.1.0 because of their changed requirement of `intl-messageformat` from `9.4.6` to `10.2.4`.

SgtPooki replied:
I wish package.json supported comments so these things could be called out more easily.

SgtPooki commented:
@lidel have those fixed locally except for the main dag text appearing differently. working on that now

SgtPooki commented:
@lidel or @hacdias whichever of you can help with this.

i've got the explore screen working properly, but I cannot for the life of me figure out how to get the data from the CID you provided, so that I can add it to the kubo node running in e2e tests, so that the explore screen will load it and run properly. I tried doing it the same way as in the other newer e2e tests in explore.test.js and it will not load without first adding it. (it works fine in webui, outside of e2e tests)

the main test i'm having trouble with is https://github.com/ipfs/ipfs-webui/pull/2098/commits/81f828d44646cf902a18a99bc2fa9fdfdcad8c75 and I believe all I need to do is get the following to work:

```js
    const cidInstance = CID.parse('bafyreihnpl7ami7esahkfdnemm6idx4r2n6u3apmtcrxlqwuapgjsciihy')
    
    // what bytes are the actual bytes for the content? how do I get the content of the CID, so that I can add it to the ipfs-backend.js node?
    const addResult = await ipfs.add(cidInstance.multihash.bytes, { cid })

```

I need `addResult.cid.toString()` to ====== `cidInstance.toString()`

SgtPooki commented:
@lidel I addressed your comments, but using some of those examples are too slow for the e2e tests so I created some different CIDs that we can use to test that things are loading properly.

lidel reviewed (changes requested):
_(mind this is late friday, so maybe i am testing wrong cid, but...)_

I tested  `bafybeifj4cifkrs5msfcfqqrbovoxtptdgmgt3ce6nvbwitfzy6gvqekl4` from fleek preview on this PR and examples are still broken, but one level deeper :)

- unable to enter into unixfs subdiretories in both examples:
  - http://bafybeifj4cifkrs5msfcfqqrbovoxtptdgmgt3ce6nvbwitfzy6gvqekl4.ipfs.localhost:8080/#/explore/QmSnuWmxptJZdLJpKRarxBMS2Ju2oANVrgbr2xWbie9b2D → click on `albums` 
  - http://bafybeifj4cifkrs5msfcfqqrbovoxtptdgmgt3ce6nvbwitfzy6gvqekl4.ipfs.localhost:8080/#/explore/QmdmQXB2mzChmMeKY47C43LxUdg1NDJ5MWcKMKxDu7RgQm/ → click on any subdir
  - both produce  ![Screenshot 2023-02-18 at 02-04-23 Exploring IPLD](https://user-images.githubusercontent.com/157609/219823234-22653577-c8c9-492d-834b-fc75621a4dce.png)
- dag-cbor has similar problem, cant visit children nodes:
  - http://bafybeifj4cifkrs5msfcfqqrbovoxtptdgmgt3ce6nvbwitfzy6gvqekl4.ipfs.localhost:8080/#/explore/bafyreihnpl7ami7esahkfdnemm6idx4r2n6u3apmtcrxlqwuapgjsciihy → click on `links/0` → `Error: No resolver found for codec "undefined"`

SgtPooki commented:
I'm able to reproduce locally with:

* http://localhost:3000/#/explore/QmSnuWmxptJZdLJpKRarxBMS2Ju2oANVrgbr2xWbie9b2D/albums
* http://localhost:3000/#/explore/QmdmQXB2mzChmMeKY47C43LxUdg1NDJ5MWcKMKxDu7RgQm/10%20-%20Pi%20Equals
* http://localhost:3000/#/explore/bafyreihnpl7ami7esahkfdnemm6idx4r2n6u3apmtcrxlqwuapgjsciihy/links/0

SgtPooki commented:
got repro of the failure with XKCD in e2e test at https://github.com/ipfs/ipfs-webui/pull/2098/commits/d542feac9e19e7dbde315d84cd96910222450e92

SgtPooki commented:
@lidel I fixed the issue you brought up, and then found that the next level was not traversable, and fixed that.. but the test still only tests one level deep. i'm updating the test to walk as far down as possible now.

SgtPooki commented:
@lidel @whizzzkid This is ready for review

SgtPooki commented on src/components/address/Address.js:
```diff
@@ -4,10 +4,7 @@ import Multiaddr from 'multiaddr'
-  value = value.buffer || value
-
-  const ma = Multiaddr(value)
+  const ma = Multiaddr(value.toString())
```
yea in a subsequent PR for sure

lidel commented on test/e2e/explore.test.js:
```diff
@@ -1,36 +1,223 @@
+ * @param {string} param0.type
+ * @returns {Promise<boolean>} true if a child was found and clicked
+ */
+async function clickFirstExploreChild ({ page, type }) {
```
@SgtPooki I am afraid this  is too naive –  for Apollo archives it hits the README.txt and never goes deeper:

- :green_circle: http://localhost:3000/#/explore/QmSnuWmxptJZdLJpKRarxBMS2Ju2oANVrgbr2xWbie9b2D/README.txt

We have hardcoded CIDs of examples, we need to add more explicit test than parent dir.

Just open a hardcoded, deterministic path that goes deep within directory tree.
Use :red_circle:  I mentioned.

SgtPooki replied:
> (not sure if needed, but in case you see test being slow/failing on CID) we can ipfs block get specific blocks that are used in tests as static test vectors into test directory, and ipfs block put them before running E2E tests, so they work in offline mode. (you can see which blocks are needed by manually traversing dag in IPLD Explorer and looking at block requests in Network tab in devtools)

That's a useful tip. I will try that out

SgtPooki replied:
Created follow up issue for this: https://github.com/ipfs/ipfs-webui/issues/2106

lidel reviewed (changes requested):
Super close! :crossed_fingers: 

Unfortunately, we have regressions before reaching deeper leaf nodes  (deeper than one level) :crying_cat_face: 

- dag-pb 
  - :orange_circle:  http://localhost:3000/#/explore/QmSnuWmxptJZdLJpKRarxBMS2Ju2oANVrgbr2xWbie9b2D/albums/QXBvbGxvIDE0IE1hZ2F6aW5lIDY0L0xM/21076541854_939b42d8d8_o.jpg opens ok, but clicking on any of its children, like `Links/0` will error:
    - :red_circle:  http://localhost:3000/#/explore/QmSnuWmxptJZdLJpKRarxBMS2Ju2oANVrgbr2xWbie9b2D/albums/QXBvbGxvIDE0IE1hZ2F6aW5lIDY0L0xM/21076541854_939b42d8d8_o.jpg/Links/0  
      >  ![2023-03-03-015821_1194x300_scrot](https://user-images.githubusercontent.com/157609/222605043-4b0dcb1a-4ce9-49ca-a96a-21bc2b67dd82.png)

- same for dag-cbor: 
   - old version opened :information_source:  http://127.0.0.1:5001/ipfs/bafybeiequgo72mrvuml56j4gk7crewig5bavumrrzhkqbim6b3s2yqi7ty/#/explore/bafyreihnpl7ami7esahkfdnemm6idx4r2n6u3apmtcrxlqwuapgjsciihy/links/0/links/0/links/0/links/0 just fine, 
 but this PR fails 2 levels of links deep: :red_circle:  http://localhost:3000/#/explore/bafyreihnpl7ami7esahkfdnemm6idx4r2n6u3apmtcrxlqwuapgjsciihy/links/0/links/0 → fails
  >  ![2023-03-03-015821_1194x300_scrot](https://user-images.githubusercontent.com/157609/222606195-f238a92f-e598-4ace-ba25-f895ae848f75.png)

SgtPooki commented:
Since i've confirmed issues with apollo archive are fixed locally, and tests are diving fully into the XKCD archives, i'm going to merge this, though I created a follow up issue for addressing test concerns: https://github.com/ipfs/ipfs-webui/issues/2106
